text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1879 @@
+- Often proposals focused at Custom DSTs preferred to combine the
+  escape-the-sized-hierarchy part with the Custom DST machinery.
+    - e.g. `DynSized` trait with `Metadata` associated types and `size_of_val`
+      and `align_of_val` methods, or a `!Sized` pseudo-trait that you could implement.
+    - Given the acceptance of [rfcs#2580][rfc_pointer_metadata_vtable], Rust
+      doesn't seem to be trending in this direction, as the `Metadata` part of this is
+      now part of a separate `Pointee` trait.
+- Most early `DynSized` trait proposals (independent or as part of Custom DSTs)
+  would make `DynSized` a default bound mirroring `Sized`, and consequently had a
+  relaxed form `?DynSized`.
+    - Later proposals were more aware of the language team's resistance towards
+      adding new relaxed bounds and tried to avoid this.
+- Backwards compatibility concerns were the overriding reason for the rejection
+  of previous `DynSized` proposals.
+    - These can be sidestepped by avoiding having a relaxed form and by relying
+      on being a supertrait of `Sized`.
+
+The [Rationale and Alternatives](#rationale-and-alternatives) section provides
+rationale for some of the decisions made in this RFC and references the prior
+art above when those proposals made different decisions.
+
+No previous proposal captures the specific part of the design space that this
+proposal attempts to, but these proposals are the closest matches for parts of
+this proposal:
+
+- [Pre-eRFC: Let's fix DSTs][pre_erfc_fix_dsts] was the only other proposal
+  removing `Sized` bounds when a bound for another sized trait (only `DynSized`
+  in that pre-eRFC's case) was present.
+    - However, this proposal had `size_of_val` methods in its `DynSized` trait and
+      proposed a bunch of other things necessary for Custom DSTs.
+- [rfcs#2310: DynSized without ?DynSized][rfc_dynsized_without_dynsized] was
+  proposed at a similar time and was similarly focused only on making `Sized` more
+  flexible, but had a bunch of machinery for avoiding backwards incompatibility
+  that this RFC believes is unnecessary. Like this proposal, it avoided making
+  `DynSized` a default bound and avoided having a relaxed form of it.
+    - However, this proposal didn't suggest removing default `Sized` bounds in
+      the presence of other size trait bounds.
+- [Sized, DynSized, and Unsized][blog_dynsized_unsized] is very similar and a
+  major inspiration for this proposal. It has everything this proposal has except
+  for the const size traits and all the additional context an RFC needs.
+
+Some prior art referenced [rust#21974][issue_regions_too_simplistic] as a limitation
+of the type system which can result in new implicit bounds or implicit supertraits
+being infeasible for implementation reasons, but [it is believed that this is no
+longer relevant][zulip_issue_regions_too_simplistic].
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
```

Comment:
It'd make sense to add [this](https://github.com/rust-lang/rfcs/pull/3729#issuecomment-2511408845) as either an unresolved question or future possibility (if it's not already there -- I skimmed but could have missed it):

> ...You're right that `T: MyTrait` would not imply removal of the default bound, it would be `T: const Sized + MyTrait` unless you wrote `T: MyTrait + const ValueSized` (or `T: MyTrait + ?Sized` if we go that way). In [Niko's blog post](https://smallcultfollowing.com/babysteps/blog/2024/04/23/dynsized-unsized/), he suggested it could be that `T: MyTrait` implies removal of the default bound. I mention that what I have proposed [is compatible with that](https://github.com/davidtwco/rfcs/blob/sized-hierarchy/text/3729-sized-hierarchy.md#forward-compatibility-with-supertraits-implying-default-bound-removal) but don't propose it at the moment.

Reply:
I do mention this in the future possibilities here:
https://github.com/rust-lang/rfcs/blob/462d016a8f8bc00bf0ce6c10e8f52bcf0903cf87/text/3729-sized-hierarchy.md?plain=1#L1684-L1689